### PR TITLE
Feature/attachment enhance

### PR DESCRIPTION
**Pull request by chrisminett**
I've tried to break this down into logical commits, but the basic premise for the changes is this:
- There was no dedicated `Attachment` class, other than that used to allow a local file to be dynamically added to an existing `Mail` object (ie. no attachment class was used from the Factory)
- A charset was sometimes incorrectly defined on an attachment
- An attachment was sometimes not recognised as an attachment (we were only looking for _disposition_, but inline attachments won't necessarily have this)

So the overall changes are:
- Split local file handling into `AttachmentLocal`, leaving `Attachment` to be the consistent representation for an attachment, both for the Factory and dynamic usage.
- Only the `Mail` class gets a default charset, everything else needs it to be explicitly defined
  - The Factory only sets a charset for Attachments if it's defined for the content, rather than the part
- _Disposition_ is optional for attachments, but _name_ is required

The varying attachment attribute combinations (name, disposition, charset) are tested by the extended Factory tests. This was the basis for all the other changes.

### CODE REVIEW

File: src/Content/AttachmentLocal.php [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+namespace Phlib\Mail\Content;
+
+use Phlib\Mail\Exception\RuntimeException;
+
+/**
+ * Attachment class used to add local files to a Mail object
+ *
+ * @package Phlib\Mail
+ */
+class AttachmentLocal extends Attachment
```
Comment (qubit05) [inline src/Content/AttachmentLocal.php]: Would this be better called FileAttachment or AttachmentFile? AttachmentLocal is confusing how it relates to an email attachment.

What's the difference between Attachment and AttachmentLocal. Is one Inline? If so, would naming it more clearly be better?

File: src/SetTypeTrait.php [hunk 2]
```diff
@@ -0,0 +1,23 @@
+<?php
+
+namespace Phlib\Mail;
+
+/**
+ * Trait to enable the content-type to be overridden
+ *
+ * @package Phlib\Mail
+ */
+trait SetTypeTrait
```
Comment (qubit05) [inline src/SetTypeTrait.php]: Not sure about the SetTypeTrait idea. I see why you're doing it.

The 'type' property should be part of the trait.

Comment (qubit05) [inline src/SetTypeTrait.php]: Looking at it a bit further, could you confirm my thinking:

It feels like there shouldn't be a setType method. The only type you want to specify it is on the Mime class. Creating a `new Mime('something/unusual')` class with other subtypes having types defined in the class? Then no `setType` method needed.

Comment (chrisminett) [reply src/SetTypeTrait.php]: I've the behaviour to use immutable MIME types in #4 so this branch isn't responsible for changing the construct params of un-related classes. Then I can update the implementation here.

File: src/Content/AttachmentLocal.php (hunk 1 above)
Comment (chrisminett) [reply src/Content/AttachmentLocal.php]: How about `AttachmentFile` or `AttachmentFromFile`, and making the _filename_ a constructor parameter, so immutable?

This class is designed to help a developer create a new part that matches the `Attachment` "interface" (so extends `Attachment`) from a file on their local filesystem. Essentially it's a shortcut to:

``` php
$part = new Attachment(mime_content_type($filename));
$part->setContent(file_get_contents($filename));
$part->setName(basename($filename));
```

Comment (qubit05) [reply src/Content/AttachmentLocal.php]: Sounds more like a static create method than a separate class to me. You've almost said it yourself in that comment, `AttachmentFromFile` and "create a new part".

Comment (chrisminett) [reply src/Content/AttachmentLocal.php]: Yeah, so `Attachment::createFromFile($filename)` instead makes sense. Only downside is that the file content would have to be loaded on construct, whereas at the moment it's left until the call to `getContent()`. Do you think that's OK? If so I'll make the change.

Comment (qubit05) [reply src/Content/AttachmentLocal.php]: Other than the concern about large files and memory issues. I think this should be fine for now.

Comment (chrisminett) [reply src/Content/AttachmentLocal.php]: OK, all waiting on #4 then.

**Comment by chrisminett** [comment]
OK, updated.

The Trait is gone. MIME type is now set in the construct (like #4) and I've put the attachment name there too, as that's a required attribute.

The _Local_ class is gone, replaced by a static _create_ method - much cleaner.

File: src/Content/Content.php [hunk 3]
```diff
@@ -2,6 +2,8 @@
 
 namespace Phlib\Mail\Content;
 
+use Phlib\Mail\SetTypeTrait;
```
Comment (qubit05) [inline src/Content/Content.php]: Remove

File: src/Mime/Mime.php [hunk 4]
```diff
@@ -2,6 +2,8 @@
 
 namespace Phlib\Mail\Mime;
 
+use Phlib\Mail\SetTypeTrait;
```
Comment (qubit05) [inline src/Mime/Mime.php]: Remove

**Comment by qubit05** [comment]
Why aren't the tests failing with the now unknown SetTypeTrait?

**Comment by chrisminett** [comment]
Seems as though it's because they were only namespace imports

**Pull request merged**